pkg/operator/status.go

```diff
@@ -98,55 +94,87 @@ func isDeploymentStatusComplete(deploy *appsv1.Deployment) bool {
+				// If there isn't at least one replica from each deployment, Available=False
+				c.setAvailableFalse(operatorConfigCopy, reason, statusMsg)
 				return false, nil
 			}
-			c.setFailingStatus(operatorConfigCopy, "Error getting service-ca-operator managed deployments", err.Error())
+			statusMsg = fmt.Sprintf("Error getting deployment %s", dep)
+			c.setFailingTrue(operatorConfigCopy, statusMsg, err.Error())
+			// If there isn't at least one replica from each deployment, Available=False
+			c.setAvailableFalse(operatorConfigCopy, reason, statusMsg)
 			return false, err
 		}
 		if existing.DeletionTimestamp != nil {
-			glog.Infof("Deployment %s is being deleted", dep)
-			statusMsg := fmt.Sprintf("Deployment %s is being deleted", dep)
-			c.setProgressingStatus(operatorConfigCopy, statusMsg)
+			statusMsg = fmt.Sprintf("Deployment %s is being deleted", dep)
+			c.setProgressingTrue(operatorConfigCopy, reason, statusMsg)
+			// If there isn't at least one replica from each deployment, Available=False
+			c.setAvailableFalse(operatorConfigCopy, reason, statusMsg)
 			return false, nil
 		}
 		if !isDeploymentStatusAvailable(existing) {
-			glog.Infof("Deployment %s does not have available replicas", dep)
-			statusMsg := fmt.Sprintf("Deployment %s does not have available replicas", dep)
-			c.setProgressingStatus(operatorConfigCopy, statusMsg)
+			statusMsg = fmt.Sprintf("Deployment %s does not have available replicas", dep)
+			c.setProgressingTrue(operatorConfigCopy, reason, statusMsg)
+			// If there isn't at least one replica from each deployment, Available=False
+			c.setAvailableFalse(operatorConfigCopy, reason, statusMsg)
 			return false, nil
 		}
 		existingDeploymentsAndReplicas++
 
-		if !isDeploymentStatusComplete(existing) {
-			glog.Infof("The deployment %s has not completed", dep)
-			statusMsg := fmt.Sprintf("Deployment %s has not completed", dep)
-			c.setProgressingStatus(operatorConfigCopy, statusMsg)
-			return false, nil
-		}
 		if isDeploymentStatusAvailableAndUpdated(existing) {
 			glog.Infof("Deployment %s is available and updated", dep)
-			ready++
+			version_ready++
+			if isDeploymentStatusComplete(existing) {
+				glog.Infof("Deployment %s has desired replicas.", dep)
+				deployment_complete++
+			} else {
+				statusMsg = fmt.Sprintf("Deployment %s is creating replicas.", dep)
+			}
+		} else {
+			statusMsg = fmt.Sprintf("Deployment %s is updating", dep)
 		}
 	}
-	// set Available if replica instances are created.
-	// report version if all deployments are available and updated
-	if ready == len(deployments) {
-		glog.Infof("All deployments managed by service-ca-operator are available and updated")
-		c.setAvailableStatus(operatorConfigCopy)
+	// Available, Updated, and Ready to report version:
+	// Here, ready to report version and set Available=True and
+	// set Progressing=False because all deployments, replicas exist
+	// and all instances are updated, no previous deployment instances exist
```

Comment:
The returning of true to report a version or not kind of confused me as well. I think it would be cleaner for this function to have a structure where you use defer for the status/version update.

Reply:
I've updated,it's more clear now.  I'm setting the version in the syncStatus.  I'm also passing a list of deployments from the targetNamespace to the syncStatus func.  Not sure if it's better, I'm testing w/ it now.

Reply:
the comments were more for me, to make sure I knew what I was doing, i shortened them

Reply:
:)